crates/bevy_math/src/curve/mod.rs

```diff
@@ -968,19 +971,135 @@ where
+///
+/// [unit domain]: `Interval::UNIT`
+pub fn quadratic_ease_in() -> impl Curve<f32> {
+    FunctionCurve {
+        domain: Interval::UNIT,
+        f: |t: f32| t.squared(),
+        _phantom: PhantomData,
+    }
+}
+
+/// A [`Curve`] mapping the [unit interval] to itself.
+///
+/// It uses the function `f(x) = 1 - (1 - x)²`
+///
+/// [unit domain]: `Interval::UNIT`
+pub fn quadratic_ease_out() -> impl Curve<f32> {
+    FunctionCurve {
+        domain: Interval::UNIT,
+        f: |t: f32| 1.0 - (1.0 - t).squared(),
+        _phantom: PhantomData,
+    }
+}
+
+/// A [`Curve`] mapping the [unit interval] to itself.
+///
+/// It uses the function `f(x) = x² * (3 - 2x)`
+///
+/// [unit domain]: `Interval::UNIT`
+pub fn cubic_curve() -> impl Curve<f32> {
+    FunctionCurve {
+        domain: Interval::UNIT,
+        f: |t: f32| t.squared() * (3.0 - 2.0 * t),
+        _phantom: PhantomData,
+    }
+}
+
+/// A [`Curve`] mapping the [unit interval] to itself.
+///
+/// It uses the function `f(n,x) = round(x * n) / n`
+///
+/// parametrized by `n`, the number of steps
+///
+/// [unit domain]: `Interval::UNIT`
+pub fn step_curve(num_steps: usize) -> impl Curve<f32> {
+    FunctionCurve {
+        domain: Interval::UNIT,
+        f: move |t: f32| (t * num_steps as f32).round() / num_steps as f32,
+        _phantom: PhantomData,
+    }
+}
+
+/// A [`Curve`] mapping the [unit interval] to itself.
+///
+/// It uses the function `f(omega,x) = (1 - (1 - x)²)(2sin(omega * x) / omega + cos(omega * x))`
+///
+/// parametrized by `omega`
+///
+/// [unit domain]: `Interval::UNIT`
+pub fn elastic_curve(omega: f32) -> impl Curve<f32> {
+    FunctionCurve {
+        domain: Interval::UNIT,
+        f: move |t: f32| {
+            (1.0 - (1.0 - t).squared()) * (2.0 * ops::sin(omega * t) / omega + ops::cos(omega * t))
+        },
+        _phantom: PhantomData,
+    }
+}
+
+/// A [`Curve`] mapping the [unit interval] to itself.
+///
+/// It uses the function
+/// `f(p0,p1,p2,p3,x) = (1 - x)³ * p0 + 3(1 - x)² * x * p1 + 3(1 - x) * x² * p2 + t³ * p3`
+///
+/// parametrized by `p0`, `p1`, `p2`, `p3`, called "control points"
+///
+/// [unit domain]: `Interval::UNIT`
+pub fn cubic_bezier_curve(control_points: [f32; 4]) -> impl Curve<f32> {
+    let [p0, p1, p2, p3] = control_points;
+    FunctionCurve {
+        domain: Interval::UNIT,
+        f: move |t: f32| {
+            (1.0 - t).cubed() * p0
+                + 3.0 * (1.0 - t).squared() * t * p1
+                + 3.0 * (1.0 - t) * t.squared() * p2
+                + t.cubed() * p3
+        },
+        _phantom: PhantomData,
+    }
+}
+
+/// Convert a `start` and `end` point of some type into a [`Curve`], sampled by interpolating
+/// between the two boundary values and under the application of an easing function.
+pub fn easing_curve<T: VectorSpace>(
+    start: T,
+    end: T,
+    easing_curve: impl Curve<f32>,
+) -> impl Curve<T> {
+    line_curve(start, end).reparametrize_by_curve(easing_curve)
```

Comment:
While I like the use of the Curves API, I feel like we could be dealing with functions mapping f32 to f32 all around here, and pass those into here.

Ofc a user can still do that by making a FunctionCurve, so I suppose there is not much lost, it does seem like a bit of overhead maybe.

Reply:
I'm not really sure what to do about this review comment. Do you have any actionable ideas?

Reply:
This was replaced by `EasingCurve` which is more explicit and has more useful functionalities (it guarantees the unit interval domain)